python/ray/experimental/workflow/workflow_manager.py

```diff
@@ -113,75 +107,45 @@ def _func(context, task_id, args, kwargs):
+                # NOTE: When calling 'ray.put', we trigger python object
+                # serialization. Under our serialization context,
+                # Workflows and ObjectRefs are separated from the arguments,
+                # leaving a placeholder object with all other python objects.
+                # Then we put the placeholder object to object store,
+                # so it won't be mutated later. This guarantees correct
+                # semantics. See "tests/test_variable_mutable.py" as
+                # an example.
+                input_placeholder: RRef = ray.put((args, kwargs))
+            return Workflow(self.func, self._run_step, input_placeholder,
+                            workflows, object_refs)
 
         self.step = _build_workflow
 
-    def _run_step(self, args, kwargs):
-        task_id = uuid.uuid4()
-        refs = self._remote_function.remote(
-            workflow_context.get_workflow_step_context(), task_id, args,
-            kwargs)
-        outputs, output_ids = wrap_step_outputs(refs)
-        return outputs
-
-
-class Workflow:
-    def __init__(self, step_func: Callable, args: Tuple,
-                 kwargs: Dict[str, Any]):
-        self.args: List[RRef] = []
-        self.kwargs: Dict[str, RRef] = {}
-        # NOTE: we must serialize the inputs here, so they cannot be
-        # mutable later.
-        for arg in args:
-            if isinstance(arg, (ray.ObjectRef, Workflow)):
-                self.args.append(arg)
-            else:
-                self.args.append(ray.put(arg))
-        for k, v in kwargs:
-            if isinstance(v, (ray.ObjectRef, Workflow)):
-                self.kwargs[k] = v
-            else:
-                self.kwargs[k] = ray.put(v)
-        self.step_func = step_func
-        self._executed = False
-        self._output: Optional[WorkflowOutputType] = None
-
-    @property
-    def executed(self) -> bool:
-        return self._executed
-
-    @property
-    def output(self):
-        if not self._executed:
-            raise Exception("The workflow has not been executed.")
-        return self._output
-
-    def execute(self) -> WorkflowOutputType:
-        """
-        Trigger workflow execution recursively.
-        """
-        if self.executed:
-            return self._output
-        input_args = []
-        input_kwargs = {}
-        for a in self.args:
-            if isinstance(a, Workflow):
-                input_args.append(a.execute())
-            else:
-                input_args.append(a)
-        for k, v in self.kwargs.items():
-            if isinstance(v, Workflow):
-                input_kwargs[k] = v.execute()
-            else:
-                input_kwargs[k] = v
-        output = self.step_func(input_args, input_kwargs)
-        if not isinstance(output, WorkflowOutputType):
-            raise TypeError("Unexpected return type of the workflow.")
-        return output
+    def _run_step(self, step_id: StepID,
```

Comment:
Should the `_run_step` and also the `remote_function` go to Workflow init? They are more related to how to execute a workflow instead of how to construct a workflow. So put it there will make the code easier to understand.

Reply:
I think `remote_function` and the actual function execution should stay in `WorkflowStepFunction`. This is because 
* Logically `WorkflowStepFunction` is wrapped around the original function, it represents the original function instead of the workflow. Workflow is only an interface to trigger execute, saving inputs and DAG dependencies.
* Physically we should export the remote function upon `workflow.step`, so other worker in the Ray cluster can access this remote function earlier; more importantly, it prevents exporting the same function again and again if we generate the workflow multiple times from the same workflow step function.